File: Contest/Multi2-2013-7-25/1004/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.StreamTokenizer;

public class Main{
    int maxn = 50010;

    class SegTree {
        class SegNode {
            int sum, val;
            boolean flag;
            int left, right;
            SegNode(int l, int r) {
                left = l;
                right = r;
            }
            int mid() {
                return (left + right) >> 1;
            }
            int length() {
                return right - left + 1;
            }
        }

        SegNode tree[] = new SegNode[maxn * 4];
        void init(int idx, int left, int right) {
            tree[idx] = new SegNode(left, right);
            if (left == right)
                return;
            int mid = tree[idx].mid();
            init(idx << 1, left, mid);
            init(idx << 1 | 1, mid + 1, right);
            push_up(idx);
        }

        void update(int idx, int left, int right, int v) {
            push_down(idx);
            if (left <= tree[idx].left && right >= tree[idx].right) {
                tree[idx].val = v;
                tree[idx].flag = true;
                return;
            }
            int mid = tree[idx].mid();
            if (left <= mid)
                update(idx << 1, left, right, v);
            if (mid < right)
                update(idx << 1 | 1, left, right, v);
            push_up(idx);
        }

        int query(int idx, int left, int right) {
            push_down(idx);
            if (left == tree[idx].left && right == tree[idx].right) {
                return tree[idx].sum;
            }
            int mid = tree[idx].mid();
            if (right <= mid)
                return query(idx << 1, left, right);
            else if (left > mid)
                return query(idx << 1 | 1, left, right);
            else
                return query(idx << 1, left, mid)
                        + query(idx << 1 | 1, mid + 1, right);
        }

        void push_down(int idx) {
            if (tree[idx].flag) {
                tree[idx].flag = false;
                int k = tree[idx].val;
                tree[idx].sum = k * tree[idx].length();
                tree[idx].val = 0;
                if (tree[idx].left != tree[idx].right) {
                    int t = idx << 1;
                    tree[t].flag = true;
                    tree[t].val = k;
                    t++;
                    tree[t].flag = true;
                    tree[t].val = k;
                }
            }
        }

        void push_up(int idx) {
            if (tree[idx].left == tree[idx].right)
                return;
            push_down(idx << 1);
            push_down(idx << 1 | 1);
            tree[idx].sum = tree[idx << 1].sum + tree[idx << 1 | 1].sum;
        }
    }

    SegTree tree = new SegTree();
    int resL, resR, n;

    boolean bin(int k, int f) {
        if (tree.query(1, k, n) == n - k + 1 || f == 0)
            return false;
        int left = k, right = n;
        while (left <= right) {
            int mid = (left + right) >> 1;
            if (tree.query(1, k, mid) < mid - k + 1) {
                resL = mid;
                right = mid - 1;
            } else
                left = mid + 1;
        }
        left = k;
        right = n;
        if (tree.query(1, k, n)+f<=n-k+1) {
            while (left <= right) {
                int mid = (left + right) >> 1;
                if (tree.query(1, k, mid) + f <= mid - k + 1) {
                    resR = mid;
                    right = mid - 1;
                } else
                    left = mid + 1;
            }
        }
        else
        {
            while (left <= right) {
                int mid = (left + right) >> 1;
                if (tree.query(1, mid,n)<n-mid+1) {
                    resR = mid;
                    left = mid + 1;
                } else
                    right=mid-1;
            }
        }
        tree.update(1, resL, resR, 1);
        resL--;
        resR--;
        return true;
    }

    void run() throws Exception {
        int cas = nextInt();
        while (cas-- > 0) {
            n = nextInt();
            int m = nextInt();
            tree.init(1, 1, n);
            while (m-- > 0) {
                if (nextInt() == 1) {
                    int k = nextInt() + 1;
                    int f = nextInt();
                    if (bin(k, f))
                        out.write(resL + " " + resR + "\n");
                    else
                        out.write("Can not put any one.\n");
                } else {
                    int a = nextInt() + 1;
                    int b = nextInt() + 1;
                    out.write(tree.query(1, a, b) + "\n");
                    tree.update(1, a, b, 0);
                }
            }
            out.write("\n");
        }
        out.close();
    }

    StreamTokenizer in = new StreamTokenizer(new BufferedReader(
            new InputStreamReader(System.in)));
    BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));

    int nextInt() throws IOException {
        in.nextToken();
        return (int) in.nval;
    }

    public static void main(String[] args) throws Exception {
        try{
        new Main().run();
        }
        catch(Exception e)
        {
            while(true);
        }
    }

}

File: Contest/Multi2-2013-7-25/1004/main.cpp
#include <map>
#include <set>
#include <list>
#include <cmath>
#include <queue>
#include <stack>
#include <bitset>
#include <vector>
#include <cstdio>
#include <string>
#include <sstream>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <algorithm>
using namespace std;
#define PB push_back
#define MP make_pair
#define SZ(v) ((int)(v).size())
#define abs(x) ((x) > 0 ? (x) : -(x))
#define FOREACH(e,x) for(__typeof(x.begin()) e=x.begin();e!=x.end();++e)
typedef long long LL;

const int MAXN = 50005;
struct SegTree {

	int sum[MAXN << 2], flag[MAXN << 2];

	int lson(int rt) {return rt << 1; }
	int rson(int rt) {return rt << 1 | 1; }

	void pushUp(int rt) {
		sum[rt] = sum[lson(rt)] + sum[rson(rt)];
	}

	void pushDown(int rt, int m) {
		if (flag[rt] != -1) {
			flag[lson(rt)] = flag[rt];
			flag[rson(rt)] = flag[rt];
			sum[lson(rt)] = (m - (m >> 1)) * flag[rt];
			sum[rson(rt)] = (m >> 1) * flag[rt];
			flag[rt] = -1;
		}
	}

	void build(int l, int r, int rt) {
		flag[rt] = -1;
		if (l == r) {
			sum[rt] = 0;
			return;
		}
		int mid = (l + r) >> 1;
		build(l, mid, lson(rt));
		build(mid + 1, r, rson(rt));
		pushUp(rt);
	}

	int query(int L, int R, int l, int r, int rt) {
		if (L <= l && r <= R) {
			return sum[rt];
		}
		pushDown(rt, r - l + 1);
		int mid = (l + r) >> 1;
		int ret = 0;
		if (L <= mid) ret += query(L, R, l, mid, lson(rt));
		if (R > mid) ret += query(L, R, mid + 1, r, rson(rt));
		return ret;
	}

	void update(int L, int R, int c, int l, int r, int rt) {
		if (L <= l && r <= R) {
			flag[rt] = c;
			sum[rt] = c * (r - l + 1);
			return;
		}
		pushDown(rt, r - l + 1);
		int mid = (l + r) >> 1;
		if (L <= mid) update(L, R, c, l, mid, lson(rt));
		if (R > mid) update(L, R, c, mid + 1, r, rson(rt));
		pushUp(rt);
	}
}tree;

int main(int argc, char const *argv[])
{
	int Test;
	scanf("%d", &Test);
	while (Test--) {
		int n, m;
		scanf("%d%d", &n, &m);
		tree.build(0, n - 1, 1);
		while (m--) {
			int op, a, b;
			scanf("%d%d%d", &op, &a, &b);
			if (op == 1) {
				int retl, retr;
				int lo = a - 1, hi = n;
				while (lo + 1 < hi) {
					int mid = (lo + hi) >> 1;
					if (tree.query(a, mid, 0, n - 1, 1) < mid - a + 1)
						hi = mid;
					else lo = mid;
				}
				retl = hi;

				if (retl >= n || b == 0) {
					puts("Can not put any one.");
				} else {
					lo = a, hi = n;
					while (lo + 1 < hi) {
						int mid = (lo + hi) >> 1;
						if (mid - a + 1 - tree.query(a, mid, 0, n - 1, 1) <= b)
							lo = mid;
						else hi = mid;
					}

					int v = lo;
					lo = a, hi = v + 1;
					while (lo + 1 < hi) {
						int mid = (lo + hi) >> 1;
						if (tree.query(mid, v, 0, n - 1, 1) < v - mid + 1)
							lo = mid;
						else hi = mid;
					}
					retr = lo;

					printf("%d %d\n", retl, retr);
					tree.update(retl, retr, 1, 0, n - 1, 1);
				}
			} else {
				int ret = tree.query(a, b, 0, n - 1, 1);
				printf("%d\n", ret);
				tree.update(a, b, 0, 0, n - 1, 1);
			}
		}
		printf("\n");
	}
	return 0;
}